docs/arch/ArchIndex.md

```diff
@@ -0,0 +1,249 @@
+# Snowfakery Architecture
+
+Note: This document covers the Snowfakery interpreter, and not the Snowfakery CCI task. The Snowfakery CCI task will be documented in its repo.
+
+The Snowfakery interpreter reads a recipe, translates it into internal data structures, executes it and then generates recipe output by sending output rows to an "output stream". The output stream formats them into a concrete formats like SQL, JSON, CSV etc.
+
+![Snowfakery Dataflow](flowdiagram.png "Snowfakery Dataflow")
+
+Obviously, Snowfakery architecture will be easier to understand in the context of the language itself, so understanding the syntax is a good first step.
+
+
+
+## Levels of Looping
+
+Snowfakery recipes are designed to be evaluated over and over again, top to bottom. Each run-through is called
+an _iteration_.
+
+Snowfakery is designed to be embedded into applications -- like CCI -- which use a 'start-stop-start-stop' pattern. We call the starts and stops _continuations_.
+
+This is useful for generating chunks of data called _portions_, and then handing off those portions to the parent application to let it do the next step.
+
+Here is the overall pattern:
+
+| CumulusCI     | Snowfakery  | Data Loader  |
+| ------------- |-------------| -------------|
+| Generate Data | Start       | Wait         |
+| Load Data     | Stop        | Start        |
+| Generate Data | Start       | Stop         |
+| Load Data     | Stop        | Start        |
+| Generate Data | Start       | Stop         |
+| Load Data     | Finish      | Start        |
+| Load Data     | Finished    | Finish       |
+
+Note that every time you Start and Stop Snowfakery, you generate a whole new Interpreter object, which re-reads the recipe. In some contexts, the new Intepreter object may be in a different process or (theoretically) on a different computer altogether.
+
+Of course this does raise the question of why you could not do the two in parallel, and the
+modern Snowfakery task does do that, but the continuation mechanism is still used for the
+following reason.
+
+Consider this recipe:
+
+```yaml
+- object: Account
+  nickname: Tata
+  just_once: True
+  fields:
+    name: Tata
+
+- object: Employee
+  fields:
+    reference: Account
+```
+
+Imagine if we needed to run this Snowfakery recipe 10,000,000 times in 12 simultaneous processes. Even so, there is one execution of the recipe which is special and that's the= first one, where the Tata Account is created.
+
+So Snowfakery would run it once snapshot the "continuation state" and then fan that continuation state out to the 12 simultaneous processes.
+
+When reading Snowfakery code, you must always think about the lifetime of each data structure:
+
+* Will it survive for a single iteration, like local variables? We call these Transients.
```

Comment:
Should we define the terms first before introducing how to find them/what to look for to establish a foundation to build on?

Reply:
I don't totally follow the suggestion @jofsky

Iteration is defined on line 16.
Continuation is define on line 18.

Did you mean these terms or others?

Reply:
I meant more the more explicit definitions with headers but I understand the desired progression. Maybe we should order the sections in the order of which they were brought up?